menuconfig.py

```diff
@@ -165,100 +165,276 @@
 
-    # A_BOLD tends to produce faint and hard-to-read text on the Windows
-    # console, especially with the old color scheme, before the introduction of
-    # https://blogs.msdn.microsoft.com/commandline/2017/08/02/updating-the-windows-console-colors/
-    BOLD = curses.A_NORMAL if _IS_WINDOWS else curses.A_BOLD
+        if not x in style:
+            raise StyleError("Unkown style keyword: {}".format(x))
 
-    # Default styling. Themes can override these settings below.
+        style[x] = sdef[x]
 
-    # Top row in the main display, with the menu path
-    PATH_STYLE               = (curses.COLOR_BLACK, curses.COLOR_WHITE,  BOLD                              )
+    return style
 
-    # Separator lines between windows. Also used for the top line in the symbol
-    # information dialog.
-    SEPARATOR_STYLE          = (curses.COLOR_BLACK, curses.COLOR_YELLOW, BOLD,            curses.A_STANDOUT)
+def style_add(name, sdef):
+    global __STYLES
+    style = style_expand(sdef)
+    __STYLES[name] = style
 
-    # List of items, e.g. the main display
-    LIST_STYLE               = (curses.COLOR_BLACK, curses.COLOR_WHITE,  curses.A_NORMAL                   )
+def style_from_string(sstr):
+    style = {}
 
-    # Style for the selected item
-    LIST_SEL_STYLE           = (curses.COLOR_WHITE, curses.COLOR_BLUE,   BOLD,            curses.A_STANDOUT)
+    for sline in sstr.split(" "):
+        sline.strip(" ")
 
-    # Like _LIST_(SEL_)STYLE, for invisible items. Used in show-all mode.
-    LIST_INVISIBLE_STYLE     = (curses.COLOR_RED,   curses.COLOR_WHITE,  curses.A_NORMAL, BOLD             )
-    LIST_INVISIBLE_SEL_STYLE = (curses.COLOR_RED,   curses.COLOR_BLUE,   BOLD,            curses.A_STANDOUT)
+        # Skip empty entries
+        if len(sline) == 0: continue
 
-    # Help text windows at the bottom of various fullscreen dialogs
-    HELP_STYLE               = PATH_STYLE
+        (skey, sval) = sline.split("=")
 
-    # Frame around dialog boxes
-    DIALOG_FRAME_STYLE       = SEPARATOR_STYLE
+        style[skey] = sval
 
-    # Body of dialog boxes
-    DIALOG_BODY_STYLE        = (curses.COLOR_WHITE, curses.COLOR_BLACK,  curses.A_NORMAL                   )
+    return style
 
-    # Edit box in pop-up dialogs
-    DIALOG_EDIT_STYLE        = (curses.COLOR_WHITE, curses.COLOR_BLUE,   curses.A_NORMAL, curses.A_STANDOUT)
+def style_from_kconfig():
+    global _kconf
 
-    # Edit box in jump-to dialog
-    JUMP_TO_EDIT_STYLE       = (curses.COLOR_WHITE, curses.COLOR_BLUE,   curses.A_NORMAL,                  )
+    kstyle = ""
+    for kfile in _kconf.kconfig_filenames:
+        with open(kfile) as kf:
+            for kl in kf.readlines():
+                kl = kl.rstrip("\n")
+                #m = re.match("^\s*#\s+$style:\s+(a-z0-9=_\s)+$", kl)
+                m = re.match("^\s*#\s+\$style:\s*([a-z0-9=_\s]+)$", kl)
+                if m:
+                    kstyle += m[1] + " "
+
+    return style_from_string(kstyle)
```

Comment:
Not sure it's worthwhile to parse style settings from Kconfig files. Styling feels like something that should be up to the user, not to the Kconfig authors.

I think it would make more sense to just look at the top-level Kconfig file too. A single menu can span multiple Kconfig files, so it wouldn't make sense to have per-Kconfig-file settings either. Parsing through all Kconfig files is pretty expensive too (at least 0.17 seconds with Linux 4.18 on my machine).

I'll take a look at the rest of the styling code a bit later.